doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,491 @@
+- Page token is (or can be) immutable, aiding thread-safety.
+- Using the page token to determine if more pages are available (e.g.
+  a `None` token means exhausted) allows the server to override the
+  page size, opening up an avenue for optimisation.
+
+#### Cons
+
+- Adding more functionality (e.g. previous page) requires additional
+  sibling functions or arguments.
+- Adding metadata (e.g. total results) requires additional sibling
+  functions or arguments.
+- Function invocations to fetch subsequent pages must pass all the same
+  arguments, such that they are valid for the page token. This is
+  error-prone and/or redundant.
+
+### Option 2 - Cursor objects / iterators
+
+Cursor objects are an abstraction over some unknown internal paging
+mechanism, which may be a page-token, or some other mechanism.
+This cursor could be an object-oriented wrapper with methods, e.g.
+`.next()`, or a data structure with opaque links, e.g. containing a URL
+to the "next" page.
+
+For example, a simple paging cursor object may look like:
+```python
+class AbstractPage:
+
+    def get(self) -> List[EntityReferences]:
+        # Current page of refs.
+        ...
+
+    def next(self, pageSize: int):
+        # Get next page.
+        ...
+
+    def remaining(self) -> Union[int, None]:
+        # None means "unsupported".
+        ...
+```
+
+#### Pros
+
+- Retrieving the data can be decoupled from traversal of pages, allowing
+  for lazy network retrieval of desired pages.
+- Additional metadata can be included in the cursor (e.g. remaining
+  pages) in an extensible way.
+- A mutable, stateful cursor object maximizes opportunities for the
+  manager plugin to optimize query caches.
+
+#### Cons
+
+- Manager plugin must implement yet another interface.
+- Complexity of binding to multiple target languages.
+- A mutable, stateful cursor violates the stateless design principle of
+  interactions with manager plugins. This causes complications, e.g.
+  thread-safety.
+- An immutable cursor object (such that `.next()` returns a new cursor
+  object), could increase optimisation complexity for the manager
+  implementation, e.g. managing query cache sharing, lifetime and
+  updates across multiple "live" cursors.
+
+## Considering batching
+
+Additional considerations arise when considering how batching can be
+handled algongside paging. There are three broad strategies
+
+- Split batching from paging, using a different API method for each.
+- Page through the results from all the batched input elements in
+  lockstep, i.e. whole-batch paging.
+- Page through the results of each batched input element independently,
+  using individual sub-queries after the initial whole-batch query,
+  i.e. per-element paging.
+
+Each of these batched paging designs combines more or less well with the
+various paging options presented above.
+
+Henceforth the offset-based solution is rejected, given the lack of
+flexibility afforded to the manager's implementation.
```

Comment:
[fix] I'd be good to call out that we also reject it for other reasons too (lack of stablity, potential overhead of record counting)

Reply:
Updated the text in ac493901